### PR TITLE
Added osu!track handler. Thanks Ameo <3

### PR DESCRIPTION
**Pull request by omkelderman**
OsuTrackHandler.java line 28-51 is pretty much a direct copy of RecommendHandler.java.

Maybe an idea to come up with a more generic solution? :stuck_out_tongue:

Shall I do that in this PR or is that something we can fix later?

TODO:
* [x] resolve above mentioned dup
* [x] make findbugs happy

### CODE REVIEW

File: tillerinobot/src/main/java/tillerino/tillerinobot/osutrack/UpdateResult.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    @SerializedName("levelup")
+    private boolean levelUp;
+
+    private boolean first;
+
+    private boolean exists;
+
+    @SerializedName("newhs")
+    private Highscore[] newHighscores;
```
Comment (Tillerino) [inline tillerinobot/src/main/java/tillerino/tillerinobot/osutrack/UpdateResult.java]: List didn't work here?

File: tillerinobot/src/main/java/tillerino/tillerinobot/osutrack/Downloader.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import java.io.IOException;
+import java.net.URL;
+
+public class Downloader {
+    //  __________________________________________________________________________________________
+    // /\                                                                                         \
+    // \_|    If you're reading this and thinking, hey, I wanna use that osutrack API as well:    |
+    //   |    --> PLEASE ask FIRST for permission from ameo (https://ameobea.me/osutrack/) <--    |
+    //   |   _____________________________________________________________________________________|_
+    //    \_/_______________________________________________________________________________________/
```
Comment (Tillerino) [inline tillerinobot/src/main/java/tillerino/tillerinobot/osutrack/Downloader.java]: That is fancy.

File: tillerinobot/src/main/java/tillerino/tillerinobot/osutrack/Downloader.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package tillerino.tillerinobot.osutrack;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.io.IOException;
+import java.net.URL;
+
+public class Downloader {
```
Comment (Tillerino) [inline tillerinobot/src/main/java/tillerino/tillerinobot/osutrack/Downloader.java]: Can we change this name? I know, let him who is without sin and all, but this is too much.

File: tillerinobot/src/main/java/tillerino/tillerinobot/IRCBot.java [hunk 4]
```diff
@@ -139,6 +129,7 @@ public IRCBot(BotBackend backend, RecommendationsManager manager,
 		commandHandlers.add(new DebugHandler(backend, resolver));
 		commandHandlers.add(new HelpHandler());
 		commandHandlers.add(new ComplaintHandler(manager));
+		commandHandlers.add(new OsuTrackHandler(new Downloader()));
```
Comment (Tillerino) [inline tillerinobot/src/main/java/tillerino/tillerinobot/IRCBot.java]: Downloader should be injected IMO.

**Comment by Tillerino** [comment]
Re: RecommendHandler copying

Sure, do it in this PR.

**Comment by omkelderman** [comment]
fuk, i need to rebase, was hoping I would be able to avoid that xd

anyways @Tillerino does this look somewhat ok?

also, that `MDC.put` thing in there right? In the `RecommendHandler` it was at the start of the `handle` method, so it would be called every single time. Now it is in the specialied handle (and so is the one in OsuTrackHandler) so it will only be called when it is actually a message for that handler. Dunno which of those two is desired behaviour...

File: tillerinobot/src/main/java/tillerino/tillerinobot/CommandHandler.java [hunk 5]
```diff
@@ -248,4 +250,50 @@ public String getChoices() {
...
+						break searchCommand;
+					}
+				}
+				return null;
+			}
+
+			return handleCommand(remaining, apiUser, userData);
+		}
+
+		public abstract Response handleCommand(String remaining, OsuApiUser apiUser, UserData userData) throws UserException,
```
Comment (Tillerino) [inline tillerinobot/src/main/java/tillerino/tillerinobot/CommandHandler.java]: I think this should be renamed to `handleArgument` to make it clear what's happening.

**Pull request merged**